client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -222,7 +222,7 @@ boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
     }
 
     @VisibleForTesting
-    String refreshToken() {
+    CompletableFuture<String> refreshToken() {
         long traceEnterId = LoggerHelpers.traceEnter(log, "refreshToken", this.scopeName, this.streamName);
         CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.get();
         if (currentRefreshFuture == null) {
```

Comment:
We can use the following snippet here. This eliminates the unnecessary recreateToken() invocation.
```
   final CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.updateAndGet(fut -> {
            final CompletableFuture<Void> tokenFuture = null;
            if (fut == null) {
                log.debug("Initiated token refresh for scope {} and stream {}", this.scopeName, this.streamName);
                return( this.recreateToken());
            } else {
                log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
                return fut;
            }
        });
```

Reply:
I notice a problem in the above suggestion (and its there with the original code too) -- its not optimizing on making the call to controller.
If two concurrent threads call refreshToken, we could have two calls made to controller to get delegation token and then one of them will race to win and set the token. 
we could do better than that by synchronizing here.

Reply:
We wanted to avoid the use of explicit locking here, as discussed somewhere in [the PR](https://github.com/pravega/pravega/pull/4215) that implemented this. While we could optimize for avoiding multiple calls, which in turn can occur under rare circumstances, explicit locking perhaps would induce additional costs. Besides the suggested optimization is not necessary for fixing the specific issue at hand. I am thinking, I should leave it as it right now.

Reply:
Using the `updateAndGet` operation would simplify the code but the unary operator would return the existing `fut` in the else condition, and the atomic update shall occur even when it is not needed.

Reply:
I've created a new enhancement issue to track/implement these suggestions: https://github.com/pravega/pravega/issues/4358, FYI.